software/main.py

```diff
@@ -0,0 +1,116 @@
+import image_processor
+import camera
+import motion
+import cv2
+import time
+import numpy as np
+from enum import Enum
+
+class State(Enum):
+    SEARCH_BALL = 1
+    DRIVE_TO_BALL = 2
+    FIND_BASKET = 3
+    THROW_BALL = 4
+
+def main_loop():
+    debug = False# if set to false wont show camera
+    ball_to_right = False
+    
+    #camera instance for normal web cameras
+    #cam = camera.OpenCVCamera(id = 2)
+    # camera instance for realsense cameras
+    cam = camera.RealsenseCamera(exposure = 100, depth_enabled=False)
+    
+    processor = image_processor.ImageProcessor(cam, debug=debug)
+
+    processor.start()
+
+    robot = motion.OmniMotionRobot()
+    robot.open()
+
+    start = time.time()
+    fps = 0
+    frame = 0
+    frame_cnt = 0
+    cam_center = cam.rgb_width / 2
+    center_offset = 20
+    cam_lower_third = cam.rgb_height/3 * 2
+
+    current_state = State.SEARCH_BALL
+    try:
+        while True:
+            processedData = processor.process_frame(aligned_depth=False)
+            if debug:
+                debug_frame = processedData.debug_frame
+
+                cv2.imshow('debug', debug_frame)
+
+                k = cv2.waitKey(1) & 0xff
+                if k == ord('q'):
+                    break
+
+            if current_state == State.FIND_BASKET:
+                robot.rotate()
+                if (processedData.basket_b.x > cam_center - center_offset) and (processedData.basket_b.x < cam_center + center_offset):
+                    current_state = State.THROW_BALL
+                continue
+            if current_state == State.THROW_BALL:
+                robot.throw()
```

Comment:
It is a little weird that the throw state handling is inside the motion module while all other game logic based parameters are here. Is there a reason? To me it seems better to handle throwing here as well, especially since you can use the camera feedback to correct aiming while throwing.